diff --git a/ha/halon/interface.h b/ha/halon/interface.h
index b98f43f1f..cece0bf3b 100644
--- a/ha/halon/interface.h
+++ b/ha/halon/interface.h
@@ -485,6 +485,8 @@ struct m0_reqh *m0_halon_interface_reqh(struct m0_halon_interface *hi);
  */
 struct m0_spiel *m0_halon_interface_spiel(struct m0_halon_interface *hi);

+M0_INTERNAL struct m0 *m0_halon_interface_m0_get(struct m0_halon_interface *hi);
+
 /** @} end of ha group */
 #endif /* __MERO_HA_HALON_INTERFACE_H__ */

diff --git a/ha/halon/interface.c b/ha/halon/interface.c
index 918b68801..b089bdcc3 100644
--- a/ha/halon/interface.c
+++ b/ha/halon/interface.c
@@ -1133,6 +1133,11 @@ struct m0_spiel *m0_halon_interface_spiel(struct m0_halon_interface *hi)
 	return working ? spiel : NULL;
 }

+M0_INTERNAL struct m0 *m0_halon_interface_m0_get(struct m0_halon_interface *hi)
+{
+	return &hi->hif_internal->hii_instance;
+}
+
 #undef M0_TRACE_SUBSYSTEM

 /** @} end of ha group */
